{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a2026f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import nibabel as nib\n",
    "import os\n",
    "import zoom.prepare as pp\n",
    "import zoom.sc_tool as sct\n",
    "\n",
    "os.chdir(\"/slurm/home/yrd/liaolab/nieshuyang/AHBA_sc\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60588914",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare and preprocess SBPs\n",
    "SBPs = [\"UTG\",\"SVG\",\"MEGtheta\",\"MOR\",\"KOR\",\"CMRO2\",\"T1T2\",\"deltaMT\"] # All the SBPs can be fetched via neuromaps\n",
    "# Assign SBPs to parcellations and perform spatial permutation test\n",
    "for SBP in SBPs:\n",
    "    SBP_parc, SBP_perm_parc = pp.process_SBP(\n",
    "        SBP=f\"=/HCPMMP/{SBP}/lh.{SBP}.fsLR.32k.func.gii\",\n",
    "        parcellation=\"/HCPMMP/lh.HCPMMP1.fsLR.32k.label.gii\",\n",
    "        atlas=\"fsLR\", density=\"32k\", hemi=\"L\", n_perm=1000, seed=123\n",
    "    )\n",
    "    SBP_parc.to_csv(f\"/HCPMMP/{SBP}/{SBP}_HCPMMP.csv\")\n",
    "    SBP_perm_parc.to_csv(f\"/HCPMMP/{SBP}/{SBP}_perm_HCPMMP.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f484e93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optimal AHBA preprocessing pipeline\n",
    "path = '/AHBA_donor_parc/genes/parcellations'\n",
    "fname = 'HCPMMP1_acpc_uncorr.nii'\n",
    "donors = ['9861', '10021', '12876', '14380', '15496', '15697']\n",
    "expr_HCPMMP, DS_HCPMMP = pp.abagen_ctx(\n",
    "    atlas={donor: nib.load(f\"{path}/S0{i+1}_H0351/{fname}\") for i,donor in enumerate(donors)},\n",
    "    atlas_info=\"/HCPMMP/HCPMMP1.0_info.csv\",\n",
    "    data_dir=\"/AHBA/microarray\",\n",
    "    donors_threshold=3, gene_stability_threshold=0.5\n",
    ")\n",
    "expr_HCPMMP.to_csv(\"/HCPMMP/expression_HCPMMP.csv\")\n",
    "DS_HCPMMP.to_csv(\"/HCPMMP/DS_HCPMMP.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53c3b31",
   "metadata": {},
   "source": [
    "We applied Seurat for the preprocessing and integration of scRNA-seq dataset, see sketch_integration.R. Since we need to further work with Python, count matrix, normalized gene expression matrix, metadata, cell embeddings and UMAP coordinates are extracted from the SeuratObject adult_ctx.rds to construct AnnData object adult_ctx.h5ad with the same information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1943a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene specificity score normalization\n",
    "adata = sc.read_h5ad(\"/scRNA/adult_ctx.h5ad\")\n",
    "sc.pp.neighbors(adata, n_pcs=30, n_neighbors=50)\n",
    "adata = sct.compute_gss(adata, n_jobs=-1)\n",
    "adata.write(\"/scRNA/adult_ctx_gss.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
